Fixes #33791 - respect global settings precedence

ezr-ondrej opened the pull request:
In the new registry we've forgot to force the global values precedence.
This forces the global values to be respected.

Lets fix nightlies

ezr-ondrej committed:
Fixes #33791 - respect global settings precedence

In the new registry we've forgot to force the global values precedence.
This forces the global values to be respected.

ekohl commented on app/presenters/setting_presenter.rb:
```diff
@@ -50,6 +50,10 @@ def readonly?
   end
 
+  def value
+    readonly? ? SETTINGS[name.to_sym] : super
```
This does two lookups in `SETTINGS`. Can't we use `fetch()`?
```suggestion
    SETTINGS.fetch(name.to_sym) { super }
```

ezr-ondrej replied:
It does two only for the global settings, so it boils down to whether is `key?` more efficient then `fetch`, but I guess it's insignificant, so using yours ;)

ekohl reviewed (commented):
This makes me wonder if we need more integration tests. For example, one that retrieves the value from the API. Unit tests don't cover these kind of things

ezr-ondrej commented:
> This makes me wonder if we need more integration tests. For example, one that retrieves the value from the API. Unit tests don't cover these kind of things

Well it depends, if then we test that the API uses the methods that we test by the unit tests then it actually does. I'm not against integration tests, but I'm against integration tests for everything.

This is weird funcionality tho, so the more tests for it the better :+1:

ekohl commented:
I now realize API was ambiguous. I meant the REST API. So more of an end-to-end test.

IMHO the challenge is to capture application behavior without blowing up the test suite execution time. That's a very fine line so I agree that we shouldn't have integration testing for everything.

ekohl reviewed (dismissed):
@tbrisker could you also take a look as well?

ekohl reviewed (changes requested):
Btw, you opened a new issue but it's really a regression in a refactor. Can't we find that issue and Refs it? That would keep things consistent.

ezr-ondrej commented:
> I meant the REST API

Done :)

ezr-ondrej commented:
> Btw, you opened a new issue but it's really a regression in a refactor. Can't we find that issue and Refs it? That would keep things consistent.

Well, it was never implemented in the new DSL, so it's not clear which of the issues introduced it. The issue appeared when we switched to the DSL, but it's not where we introduced it...
I've connected the bug with the tracker for the DSL effort, isn't it enough?

ekohl reviewed (approved):
That works for me.

tbrisker commented on app/presenters/setting_presenter.rb:
```diff
@@ -50,6 +50,10 @@ def readonly?
   end
 
+  def value
+    SETTINGS.fetch(name.to_sym) { super }
```
where does `super` go in this case? this class has no parent?

ekohl replied:
There's `attribute :value` which probably does some magic with `ActiveModel`, but it's a good point.

ezr-ondrej replied:
it includes ActiveModel::Attributes which defines the super method.
Do you think it would be more clear to call the attribute fetch ?

tbrisker replied:
Not crucial but i think that might be a bit clearer to read
```suggestion
    SETTINGS.fetch(name.to_sym, attribute(:value))
```

tbrisker replied:
opened https://github.com/theforeman/foreman/pull/8883 to propose this change, feel free to close it if you disagree

ehelms commented:
I can confirm this solved the nightly install breakage for me when running it.

tbrisker reviewed (approved):
Merging now to unblock nightlies, we can followup with changing the fetch style in another pr.

ares commented:
no need to CP, because 3.0 didn't use any setting defined through the DSL
